df_chat/tests/test_message_endpoint.py

```diff
@@ -17,15 +22,141 @@ def test_message_creation_success(self):
+
+        return user, room
+
+    def test_filter_message_by_keyword(self):
+        """
+        Test-case for filtering messages by the `Message` model's `body` field.
+        """
+        _, room = self.test_message_creation_success()
+        messages_list_endpoint = reverse(
+            "rooms-messages-list", kwargs={"room_pk": room.pk}
+        )
+
+        for body in [
+            "Test message 1234.",
+            "Test message 1234",
+            "Test",
+            "test",
+            "Message",
+            "message",
+            "1234",
+        ]:
+            messages_list_response = self.client.get(
+                messages_list_endpoint,
+                {"body": body},
+            )
+            # Assert the correct HTTP status code for response.
+            self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
+            # Assert the response body contains the keyword.
+            self.assertIn(
+                body.lower(), messages_list_response.json()[0]["body"].lower()
+            )
+
+        for body in [
+            "somethingrandom",
+            "messages",
+            "124",
+        ]:
+            messages_list_response = self.client.get(
+                messages_list_endpoint,
+                {"body": body},
+            )
+            # Assert the correct HTTP status code for response.
+            self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
+            # Assert the response is an empty list.
+            self.assertEqual(messages_list_response.json(), [])
+
+    def test_filter_message_by_username(self):
+        """
+        Test-case for filtering messages by the `Message` model's related `User`'s `username` field.
+        """
+        user, room = self.test_message_creation_success()
+        messages_list_endpoint = reverse(
+            "rooms-messages-list", kwargs={"room_pk": room.pk}
+        )
+
+        messages_list_response = self.client.get(
+            messages_list_endpoint,
+            {"username": user.username},
+        )
+        # Assert the correct HTTP status code for response.
+        self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
+        # Assert the message `room_user_id` matches the `id` field of the `RoomUser` object
+        #   corresponding to the user.
+        room_user_obj = RoomUser.objects.get(user__id=user.id)
+        self.assertEqual(
+            messages_list_response.json()[0]["room_user_id"], room_user_obj.id
+        )
+
+        messages_list_response = self.client.get(
+            messages_list_endpoint,
+            {"username": "invalid_username"},
+        )
+        # Assert the correct HTTP status code for response.
+        self.assertEqual(messages_list_response.status_code, http.HTTPStatus.OK)
+        # Assert the response is an empty list.
+        self.assertEqual(messages_list_response.json(), [])
+
+    def test_filter_message_by_date_range(self):
```

Comment:
Please test the edge cases of the filter. 
- The sample data should include messages that fall within a range as well as out of the range of the filter.
- Then in each test, make sure that the response only returns the data from within the requested range.

In this case, I see that only one message is used as a sample data.  Please increase the sample size to atleast 3 messages such that 2 of them fall within the requested range.

Reply:
Done! The new scenarios in the `TestMessageEndpoint.test_filter_message_by_date_range` should cover all of these. It creates multiple messages in-between multiple timestamps and checks the API results against all of them.

Also, since messages created via to back-to-back calls to `TestMessageEndpoint.test_message_creation_success` do not persist (due to database rollback after each test-case), messages are created manually in  `TestMessageEndpoint.test_filter_message_by_date_range`.

An alternative is to write a custom test runner that can exclude single test cases from database rollback, but this needs approval.